### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by andela-akolawole**
**Are you submitting to `development`?**
Yes

### Acceptance Criteria
As an admin, I should see a notification for new orders.As a registered user, I should see notifications for canceled orders and funds transfer.

As a user:
- GIVEN, my order is canceled, I should receive a notification for my canceled order in the app and SMS.
- GIVEN, a payment has been made to my wallet, I should receive a notification.
- WHEN I click on the notification in the app, I should be redirected to the order.
- THEN, I should be able to clear the notifications.

**DEV NOTES**
- Make use of the order workflow concept.
- Make use of an SMS gateway for SMS notifications.
For sending notifications use this meteor method
```js
const userId = "ddzuN2YPvgvx7rJS5"; // The receiver
const type = "orderDelivered"; // The notification type
const url = "/order"; //The route to follow when a notiifcation is clicked
const sms = true; // set to false if you do not want to send an sms notification
const details = "details about the notiifcation"; // The details of the notification

Meteor.call("notification/send", userId, type, url, sms, details);
```
**Note**
The meteor method argument "type" accepts strings like:

String | Message
---------| -----------
orderCancelled | Your order was canceled.
newOrder | Your order is being processed
topWallet |Your wallet has been credited
forAdmin | You have a new order
orderDelivered | Your order has been delivered
orderAccepted | Your order has been accepted
orderShipped | Your order has been shipped

**DESIGN NOTES**
The notification should be seen at the navbar with a bell icon for identification.Once clicked, a drop down should display a maximum of five notifications with a view all.

- [x] Only contains code directly related to the issue
- [ ] Has tests.
- [x] Has docs.
- [ ] Passes all tests
- [x] Has been linted and follows the style guide

### CODE REVIEW

**Comment by andela-akolawole** [comment]
@zenweasel This is an updated version of the former PR.

**Comment by andela-akolawole** [comment]
@aaronjudd I need this PR to be reviewed.

**Comment by zenweasel** [comment]
@andela-akolawole We are closed this week for holiday break. But when we are back I will take a look at it right away.

**Comment by andela-akolawole** [comment]
@zenweasel Thanks

File: imports/plugins/included/sms/server/methods/sms.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * sms/saveSettings
+   * @summary This save the sms provider settings
+   * @param {Object} settings - settings
+   * @return {object} returns result
+   */
+  "sms/saveSettings": (settings) => {
+    check(settings, Object);
+    settings.shopId = Reaction.getShopId();
+    Meteor.subscribe("SmsSettings").ready();
```
Comment (zenweasel) [inline imports/plugins/included/sms/server/methods/sms.js]: I don't think you want to call `ready` there. https://docs.meteor.com/api/pubsub.html#Subscription-ready

File: lib/collections/schemas/notifications.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+
+export const Notification = new SimpleSchema({
+  message: {
+    type: String,
+    optional: false
+  },
+  type: {
+    type: String,
+    optional: false
```
Comment (zenweasel) [inline lib/collections/schemas/notifications.js]: Shouldn't we provide what the types are?

**Comment by zenweasel** [comment]
So except for a few minor things this looks good except for the fact that I don't see where Notifications are actually being sent? Am I missing something? Shouldn't you be adding in sending notifications to the event hooks for order created etc.?

File: imports/plugins/included/notifications/server/methods/notifications.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  "notification/send": function (userId, type, url, sms, details) {
+    check(userId, String);
+    check(type, String);
+    check(sms, Boolean);
+    check(details, Match.OptionalOrNull(String));
+    check(url, String);
+
+    const values = {};
+    const types = {
+      orderCancelled: "Your order was canceled.",
```
Comment (andela-akolawole) [inline imports/plugins/included/notifications/server/methods/notifications.js]: I included i18n. Check the translations.

**Comment by andela-akolawole** [comment]
@zenweasel Updates have been made.

**Comment by andela-akolawole** [comment]
@mikemurray Fixed

**Comment by aaronjudd** [comment]
👍

**Comment by zenweasel** [comment]
I was going to suggest you use Event Hooks instead of Method Hooks but I see that a) The docs for theser are on the site and b) we don't have an event for orderCreate, etc.

So I think this is ready to go.

**Comment by zenweasel** [comment]
Approved

**Comment by andela-akolawole** [comment]
Thanks

File: imports/plugins/included/notifications/server/hooks/notification.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+const getAdminUserId = () => {
+  const admin = Accounts.find().fetch();
+  return admin[0]._id;
+};
+
+const sendNotificationToAdmin = () => {
+  const adminId = getAdminUserId();
+  const type = "forAdmin";
+  const url = "/reaction/dashboard/orders";
```
Comment (aaronjudd) [inline imports/plugins/included/notifications/server/hooks/notification.js]: we actually cannot depend on the url being /reaction/ here.  This is the shop name, slugged.  (isn't it)?

```
  const prefix = Reaction.getSlug(Reaction.getShopName().toLowerCase());
  const url = `/${prefix}/dashboard/orders`;
```

**Comment by zenweasel** [comment]
Four tests appear to be failing on this branch now.

```
I20170105-13:25:56.469(8)?   220 passing (55s)
I20170105-13:25:56.469(8)?   11 pending
I20170105-13:25:56.471(8)?   4 failing
I20170105-13:25:56.471(8)? 
I20170105-13:25:56.471(8)?   1) Add/Create cart methods cart/copyCartToOrder should throw error if cart user not current user:
I20170105-13:25:56.472(8)?      AssertionError: expected [Function: copyCartFunc] to throw 'Error' but 'TypeError: Cannot read property \'_id\' of undefined' was thrown
I20170105-13:25:56.473(8)?       at Assertion.assertThrows (packages/practicalmeteor_chai.js:2159:14)
I20170105-13:25:56.473(8)?       at Assertion.ctx.(anonymous function) [as throw] (packages/practicalmeteor_chai.js:4058:25)
I20170105-13:25:56.474(8)?       at Test.<anonymous> (server/methods/core/cart-create.app-test.js:161:7)
I20170105-13:25:56.474(8)?       at run (packages/practicalmeteor:mocha-core/server.js:56:16)
I20170105-13:25:56.474(8)? 
I20170105-13:25:56.475(8)?   2) Add/Create cart methods cart/copyCartToOrder should throw error if cart has no items:
I20170105-13:25:56.476(8)?      AssertionError: expected [Function: copyCartFunc] to throw 'Error' but 'TypeError: Cannot read property \'_id\' of undefined' was thrown
I20170105-13:25:56.484(8)?       at Assertion.assertThrows (packages/practicalmeteor_chai.js:2159:14)
I20170105-13:25:56.484(8)?       at Assertion.ctx.(anonymous function) [as throw] (packages/practicalmeteor_chai.js:4058:25)
I20170105-13:25:56.484(8)?       at Test.<anonymous> (server/methods/core/cart-create.app-test.js:184:7)
I20170105-13:25:56.485(8)?       at run (packages/practicalmeteor:mocha-core/server.js:56:16)
I20170105-13:25:56.485(8)? 
I20170105-13:25:56.486(8)?   3) Add/Create cart methods cart/copyCartToOrder should throw an error if order creation has failed:
I20170105-13:25:56.487(8)?      AssertionError: expected [Function: copyCartFunc] to throw 'Error' but 'TypeError: Cannot read property \'_id\' of undefined' was thrown
I20170105-13:25:56.487(8)?       at Assertion.assertThrows (packages/practicalmeteor_chai.js:2159:14)
I20170105-13:25:56.487(8)?       at Assertion.ctx.(anonymous function) [as throw] (packages/practicalmeteor_chai.js:4058:25)
I20170105-13:25:56.488(8)?       at Test.<anonymous> (server/methods/core/cart-create.app-test.js:197:7)
I20170105-13:25:56.489(8)?       at run (packages/practicalmeteor:mocha-core/server.js:56:16)
I20170105-13:25:56.489(8)? 
I20170105-13:25:56.489(8)?   4) Add/Create cart methods cart/copyCartToOrder should create an order:
I20170105-13:25:56.490(8)?      AssertionError: expected [Function: copyCartFunc] to throw 'Error' but 'TypeError: Cannot read property \'_id\' of undefined' was thrown
I20170105-13:25:56.490(8)?       at Assertion.assertThrows (packages/practicalmeteor_chai.js:2159:14)
I20170105-13:25:56.490(8)?       at Assertion.ctx.(anonymous function) [as throw] (packages/practicalmeteor_chai.js:4058:25)
I20170105-13:25:56.491(8)?       at Test.<anonymous> (server/methods/core/cart-create.app-test.js:216:7)
I20170105-13:25:56.491(8)?       at run (packages/practicalmeteor:mocha-core/server.js:56:16)
I20170105-13:25:56.493(8)? 
I20170105-13:25:56.493(8)? 
I20170105-13:25:56.493(8)? 
```

**Comment by andela-akolawole** [comment]
@zenweasel Are the changes I made causing the test to fail?

**Comment by zenweasel** [comment]
@andela-akolawole I believe so, yes. It seems to be tests for `copyCartToOrder`. Possibly your hooks are expecting something that's not in the test method. Couldn't tell exactly.

**Comment by zenweasel** [comment]
@andela-akolawole Which merge are you talking about?

**Comment by zenweasel** [comment]
I went back to the commit before the merge with development and the failures are there as well

**Comment by zenweasel** [comment]
I almost have the issue with the tests figured out. Will finish in the morning.

**Comment by andela-akolawole** [comment]
@zenweasel What command did you use for running rest?

**Comment by zenweasel** [comment]
`reaction test`

**Comment by zenweasel** [comment]
@andela-akolawole I hit you up on Gitter if you want to talk for a minute

**Comment by zenweasel** [comment]
@andela-akolawole Pushed commit that fixes tests. Will just retest order workflow in the morning and merge.

**Comment by andela-akolawole** [comment]
@zenweasel ok

**Comment by aaronjudd** [comment]
Tests passed, made a few more updated to check admin userId and that the addressbook had a phone number. Merged.

**Comment by andela-akolawole** [comment]
@zenweasel Could make sure the order workflow is correct?

Because I think we have an issue

```js
const getAdminUserId = () => {
  const admin = Meteor.users.findOne({
    "roles.__global_roles__": "owner"
  });
//admin would be an object not an array
  if (admin) {
    return admin[0]._id;
  }
  return false;
};
```

**Pull request merged**